src/config.c

```diff
@@ -854,30 +854,46 @@ void configSetCommand(client *c) {
 
 void configGetCommand(client *c) {
     void *replylen = addReplyDeferredLen(c);
-    int matches = 0;
+    dict *matches = dictCreate(&externalStringSetType);
     int i;
 
-    dictIterator *di = dictGetIterator(configs);
-    dictEntry *de;
-    while ((de = dictNext(di)) != NULL) {
-        standardConfig *config = dictGetVal(de);
-        for (i = 0; i < c->argc - 2; i++) {
-            robj *o = c->argv[2+i];
-            char *pattern = o->ptr;
+    for (i = 0; i < c->argc - 2; i++) {
+        robj *o = c->argv[2+i];
+        sds name = o->ptr;
 
+        /* If the string doesn't contain glob patterns, just directly
+         * look up the key in the dictionary. */
+        if (!stringContains(name, sdslen(name), "[*?")) {
+            if (dictFind(matches, name)) continue;
+            standardConfig *config = lookupConfig(name);
+
+            if (config) {
+                dictAdd(matches, name, NULL);
+                addReplyBulkCString(c, config->name);
+                addReplyBulkSds(c, config->interface.get(config->data));
+            }
+            continue;
+        }
+
+        /* Otherwise, do a match against all items in the dictionary. */
+        dictIterator *di = dictGetIterator(configs);
+        dictEntry *de;
+        while ((de = dictNext(di)) != NULL) {
+            standardConfig *config = dictGetVal(de);
             /* Note that hidden configs require an exact match (not a pattern) */
-            if (((config->flags & HIDDEN_CONFIG) && !strcasecmp(pattern, de->key)) ||
-                 (!(config->flags & HIDDEN_CONFIG) && stringmatch(pattern, de->key, 1)))
-            {
-                addReplyBulkCString(c, de->key);
+            if (config->flags & HIDDEN_CONFIG) continue;
+
+            if (dictFind(matches, config->name)) continue;
```

Comment:
if we already bothered with that `matches` dict to avoid duplicate configs, we can use this opportunity to avoid the deferred reply (has performance penalty).
i.e. add a loop at the end that users the `interface.get` after we know the size of the reply.

Reply:
Sure